state/src/replication/state_machine/recovery.rs

```diff
@@ -0,0 +1,47 @@
+//! Recover the state machine from a snapshot
+
+use common::types::tasks::{QueuedTask, TaskQueueKey};
+use openraft::SnapshotMeta;
+use tracing::info;
+use uuid::Uuid;
+
+use crate::{replication::error::ReplicationV2Error, TASK_QUEUE_TABLE};
+
+use super::StateMachine;
+
+impl StateMachine {
+    /// Check for a snapshot to recover from
+    pub(crate) async fn maybe_recover_snapshot(&mut self) -> Result<(), ReplicationV2Error> {
+        if !self.snapshot_archive_path().exists() {
+            return Ok(());
+        }
+
+        // Open the snap DB and apply a dummy snapshot
+        info!("snapshot found, recovering...");
+        let snap_db = self.open_snap_db().await?;
+        let dummy_meta = SnapshotMeta {
+            last_log_id: None,
+            last_membership: Default::default(),
+            snapshot_id: "recovery".to_string(),
+        };
+
+        self.update_from_snapshot(&dummy_meta, snap_db).await?;
+        self.clear_task_queues()
+    }
+
+    /// Clear all task queues
+    ///
+    /// We do this when recovering to prevent wallets from being blocked on a
+    /// task queue that failed
+    fn clear_task_queues(&self) -> Result<(), ReplicationV2Error> {
```

Comment:
why is this safe to do? we might recover the state machine into an inconsistent state from the snapshot. will subsequent logs apply on top of it correctly?

Reply:
i'd not say it's safe, but it's a better outcome than blocking a wallet. for one, there's no other action that is easy to do, i.e. continuing the task would be really difficult. and if we leave it blocked the wallet can't make progress.

it's not an assumption i'd like to rely on, but the startup task will look for wallet updates on-chain too.

essentially, i see this as the best of not great options in the case of a crash